File: Data Structures & Algorithms/validate-parentheses/submission-1.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.append(c)
            else:
                if len(stack) == 0:
                    return False
                cc = stack.pop()
                if self.rev(c) != cc:
                    return False
        return len(stack) == 0
    def rev(self, s: str) -> bool:
        match s:
            case ')':
                return '('
            case '}':
                return '{'
            case ']':
                return '['
            case _:
                return None

File: Data Structures & Algorithms/asteroid-collision/submission-1.ts
class Solution {
    /**
     * @param {number[]} asteroids
     * @return {number[]}
     */
    asteroidCollision(asteroids: number[]): number[] {
        let stack = [];
        for (const b of asteroids) {
            let blive = true;
            while (stack.length > 0) {
                let a = stack.pop();
                if (b < 0 && a > 0) {
                    if (Math.abs(a) > Math.abs(b)) {
                        stack.push(a);
                        blive = false;
                        break;
                    }
                    if (Math.abs(a) === Math.abs(b)) {
                        blive = false
                        break;
                    }
                } else {
                    stack.push(a);
                    break;
                }
            }
            if (blive) {
                stack.push(b);
            }
        }
        return stack;
    }
}
